### ISSUE TITLE
Enable Custom ID in release builds with Browser Integration

### ISSUE THREAD

**Issue by TomasRiker** [keepassxreboot/keepassxc#9074]
I installed the browser extension not via the Chrome Web Store, but manually (because I'm afraid a hacker might gain access to a developer's machine and upload a malicious update of the browser extension, which will automatically be deployed to all users and steal their passwords).

After installing the browser extension, I manually edited the registry key `Computer\HKEY_CURRENT_USER\SOFTWARE\Google\Chrome\NativeMessagingHosts\org.keepassxc.keepassxc_browser` to point to the `org.keepassxc.keepassxc_browser_chrome.json` file. Also, I edited that file and put the ID of my manually installed extension in there, instead of the "official" ID of the extension from the Chrome Web Store.

Whenever I save some settings (even if I didn't touch any of the settings related to browser integration), KeePassXC changes the JSON file back and puts the extension IDs from the Chrome Web Store in there, overwriting my custom ID. As a result, my extension doesn't work any more, until I perform the changes again.

KeePassXC - Version 2.7.4
Revision: 63b2394
Operating System: Windows

**Comment by varjolintu** [comment]
Usually making changes to Browser Integration settings rewrites the JSON files. Maybe that shouldn't be done if the file already exists and _"Update native messaging manifest files at startup"_ is disabled.

**Comment by TomasRiker** [comment]
> Usually making changes to Browser Integration settings rewrites the JSON files. Maybe that shouldn't be done if the file already exists and _"Update native messaging manifest files at startup"_ is disabled.

Yes! Note that it happens even when not making any changes to any settings. Simply opening the settings and clicking OK will trigger it. I think it should only rewrite the JSON files when a setting related to it has been changed.

**Comment by varjolintu** [comment]
Clicking OK = changing the settings here :) If you press Cancel, nothing is changed.

**Comment by TomasRiker** [comment]
Ideally, one would be able to configure the extension ID(s) inside the KeeStoreXC settings.
I imagine the developers of the browser extension run into a similar problem when they locally test some changes to the extension because probably it will have a different ID compared to the public version from the Chrome Web Store.

**Comment by varjolintu** [comment]
Actually that feature is implemented in the debug build, especially for the developers.
<img width="792" alt="Screenshot 2023-02-06 at 12 59 03" src="https://user-images.githubusercontent.com/24570482/216954709-60e1998e-cc44-49c6-8c20-86af2ae09a13.png">

**Comment by TomasRiker** [comment]
That's great! Maybe it should be promoted to a regular feature :)

I think installing the extension manually is a good idea (and should be made as easy as possible), especially if you're concerned that some day some attacker might publish a malicious password-stealing update to the browser extension which will then be automatically downloaded and installed by all users.

**Comment by varjolintu** [comment]
We could change that custom extension ID feature to be included in the release builds as well, if others agree with that.

**Comment by TomasRiker** [comment]
Is there progress on this issue?

**Comment by varjolintu** [comment]
@TomasRiker No, not yet.

**Comment by TomasRiker** [comment]
As far as I see, only the `#ifdef` ... `#endif` has to be removed in these places:
1. https://github.com/keepassxreboot/keepassxc/blob/a5c1298e3226a59b7430af96757565b0308da312/src/browser/BrowserSettings.h#L60
2. https://github.com/keepassxreboot/keepassxc/blob/a5c1298e3226a59b7430af96757565b0308da312/src/browser/BrowserSettings.cpp#L203

Right?

**Comment by varjolintu** [comment]
@TomasRiker Yes, but I'm tempted to change this feature to use this: https://github.com/keepassxreboot/keepassxc/pull/9406. Not sure if it's a wise choice to enable a custom ID and save it to a config file. Database(s) could hold the value instead.

**Comment by TomasRiker** [comment]
> @TomasRiker Yes, but I'm tempted to change this feature to use this: #9406. Not sure if it's a wise choice to enable a custom ID and save it to a config file. Database(s) could hold the value instead.

That would be stored unprotected (not encrypted) in the database, right? Otherwise the browser plugin would not be able to communicate with the application before the database is unlocked.

**Comment by varjolintu** [comment]
> That would be stored unprotected (not encrypted) in the database, right? Otherwise the browser plugin would not be able to communicate with the application before the database is unlocked.

True. And that's a problem with this feature. If databases have different Custom ID's, the manifest files are constantly edited, and this of course will not work properly. Enabling it globally with non-debug builds would allow any application to modify the custom ID locally (and some people are not happy about that).

**Comment by varjolintu** [comment]
> That would be stored unprotected (not encrypted) in the database, right? Otherwise the browser plugin would not be able to communicate with the application before the database is unlocked.

True. And that's a problem with this feature. If databases have different Custom ID's, the manifest files are constantly edited, and this of course will not work properly. Enabling it globally with non-debug builds would allow any application to modify the custom ID locally (and some people are not happy about that). But then again, manifest files could be also edited, so I see no harm to enable this feature.

**Comment by droidmonkey** [comment]
You could have a database setting that allows connecting to a custom extension ID (if enabled/set). Might be overkill though.

**Comment by Syuks** [comment]
In the meantime, for anyone interested, I decided to set the .json file to Read-only. Just an error popup is less troublesome than editing the .json file everytime I change any setting or forget to not press OK. So far, no issues.

**Comment by varjolintu** [comment]
> In the meantime, for anyone interested, I decided to set the .json file to Read-only. Just an error popup is less troublesome than editing the .json file everytime I change any setting or forget to not press OK. So far, no issues.

Or just.. https://github.com/keepassxreboot/keepassxc/issues/9074#issuecomment-1418875357

**Comment by Syuks** [comment]
Sorry, I don't know if I understood your suggestion correctly. If you mean disabling the setting "Update native messaging manifest files at startup", that doesn't solve the issue. I have it disabled, but the .json files are still rewritten every time I change settings or just click OK.


https://github.com/keepassxreboot/keepassxc/assets/64665272/4ca7dfcf-db99-4a25-b344-74cd639158f3
